src/Compilers/CSharp/Portable/Binder/WithExtensionParameterBinder.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Binder used to place extension parameter, if any, in scope.
+    /// </summary>
+    internal sealed class WithExtensionParameterBinder : Binder
+    {
+        private readonly NamedTypeSymbol _type;
+
+        internal WithExtensionParameterBinder(NamedTypeSymbol type, Binder next)
+            : base(next)
+        {
+            _type = type;
+        }
+
+        internal override void AddLookupSymbolsInfoInSingleBinder(LookupSymbolsInfo result, LookupOptions options, Binder originalBinder)
+        {
+            if (options.CanConsiderMembers())
+            {
+                if (_type.ExtensionParameter is { Name: not "" } parameter &&
+                    originalBinder.CanAddLookupSymbolInfo(parameter, options, result, null))
+                {
+                    result.AddSymbol(parameter, parameter.Name, 0);
+                }
+            }
+        }
+
+        internal override void LookupSymbolsInSingleBinder(
+            LookupResult result, string name, int arity, ConsList<TypeSymbol> basesBeingResolved, LookupOptions options, Binder originalBinder, bool diagnose, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+        {
+            Debug.Assert(result.IsClear);
+
+            if ((options & (LookupOptions.NamespaceAliasesOnly | LookupOptions.NamespacesOrTypesOnly)) != 0)
```

Comment:
>| LookupOptions.NamespacesOrTypesOnly [](http://example.com/codeflow?start=63&length=37)

I didn't follow, what's the reasoning for checking some options early, but letting others be handled by `CheckViability`?
For comparison, `InMethodBinder` and `WithLambdaParametersBinder` only bail out early for `NamespaceAliasesOnly`.

Reply:
> I didn't follow, what's the reasoning for checking some options early, but letting others be handled by 

Why do the work if we know that parameters do not satisfy the request. Especially, when we are binding types names in source, we want avoid having to create symbols for members, etc. This helps with avoiding cycles. In terms of implementation, I used ```WithPrimaryConstructorParametersBinder``` as a template.